impress-org/givewp
# Past donation data migration for revenue database table is compatible with currency switcher

Resolves #5381

Addon Compatibility PRs:
- https://github.com/impress-org/give-currency-switcher/pull/284
- https://github.com/impress-org/give-funds/pull/86

## Description
Currency switcher store donation amount in base currency in `_give_cs_base_amount` donation meta key. I updated data migration logic to get donation amount from `_give_cs_base_amount` donation meta and by default donation amount will be taken from `_give_donation_total`.

## Affects
In the revenue database table, we store donation amount in the smallest unit by multiply with `100`. We can not get the smallest unit of all currencies by multiply with `100` like Bahraini dinar (3 decimal), Japanese Yen (0 decimal). To fix resolve this, I added a `Give\ValueObjectsMoney` class in the core to calculate the smallest unit of any currency. I updated related code in Give Core, Fund addon, and Currency Switcher addon (pr links given above).

## Pre-review Checklist
-   [x] `@since` tag added

## Testing Instructions
- This is possible that `AddPastDonationsToRevenueTable` migration is already run on your existing install, so test this pr on a fresh install. 
- You can test this pr with or without the Currency switcher add-on, Fund addon, Recurring Donation.
- Make sure few donations exist in the database which donates with other than the base currency.
- Check `Multi Goal Process Bar` because I refactor code to use the `Money` class.
